libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpVP9Reader.java

```diff
@@ -0,0 +1,230 @@
+
+  /** Creates an instance. */
+  public RtpVP9Reader(RtpPayloadFormat payloadFormat) {
+    this.payloadFormat = payloadFormat;
+    firstReceivedTimestamp = C.TIME_UNSET;
+    startTimeOffsetUs = 0;
+    previousSequenceNumber = C.INDEX_UNSET;
+    fragmentedSampleSizeBytes = 0;
+    width = C.INDEX_UNSET;
+    height = C.INDEX_UNSET;
+    gotFirstPacketOfVP9Frame = false;
+    isKeyFrame = false;
+    isOutputFormatSet = false;
+  }
+
+  @Override
+  public void createTracks(ExtractorOutput extractorOutput, int trackId) {
+    trackOutput = extractorOutput.track(trackId, C.TRACK_TYPE_VIDEO);
+    castNonNull(trackOutput).format(payloadFormat.format);
+  }
+
+  @Override
+  public void onReceivingFirstPacket(long timestamp, int sequenceNumber) {}
+
+  @Override
+  public void consume(ParsableByteArray data, long timestamp, int sequenceNumber, boolean rtpMarker)
+      throws ParserException {
+    checkStateNotNull(trackOutput);
+
+    if (parseVP9Descriptor(data, sequenceNumber)) {
+      if (fragmentedSampleSizeBytes == 0 && gotFirstPacketOfVP9Frame) {
+        isKeyFrame = (data.peekUnsignedByte() & 0x04) == 0;
+      }
+
+      if (!isOutputFormatSet && width > 0 && height > 0) {
+        if (width != payloadFormat.format.width || height != payloadFormat.format.height) {
+          trackOutput.format(
+              payloadFormat.format.buildUpon().setWidth(width).setHeight(height).build());
+        }
+        isOutputFormatSet = true;
+      }
+
+      int fragmentSize = data.bytesLeft();
+      // Write the video sample.
+      trackOutput.sampleData(data, fragmentSize);
+      fragmentedSampleSizeBytes += fragmentSize;
+
+      if (rtpMarker) {
+        if (firstReceivedTimestamp == C.TIME_UNSET) {
+          firstReceivedTimestamp = timestamp;
+        }
+        bufferFlags = isKeyFrame ? C.BUFFER_FLAG_KEY_FRAME : 0;
+        long timeUs = toSampleUs(startTimeOffsetUs, timestamp, firstReceivedTimestamp);
+        trackOutput.sampleMetadata(
+            timeUs,
+            bufferFlags,
+            fragmentedSampleSizeBytes,
+            /* offset= */ 0,
+            /* encryptionData= */ null);
+        fragmentedSampleSizeBytes = 0;
+        gotFirstPacketOfVP9Frame = false;
+      }
+      previousSequenceNumber = sequenceNumber;
+    }
+  }
+
+  @Override
+  public void seek(long nextRtpTimestamp, long timeUs) {
+    firstReceivedTimestamp = nextRtpTimestamp;
+    fragmentedSampleSizeBytes = 0;
+    startTimeOffsetUs = timeUs;
+  }
+
+  // Internal methods.
+  private static long toSampleUs(
+      long startTimeOffsetUs, long rtpTimestamp, long firstReceivedRtpTimestamp) {
+    return startTimeOffsetUs
+        + Util.scaleLargeTimestamp(
+            (rtpTimestamp - firstReceivedRtpTimestamp),
+            /* multiplier= */ C.MICROS_PER_SECOND,
+            /* divisor= */ MEDIA_CLOCK_FREQUENCY);
+  }
+
+  private static boolean parseVP9Descriptor(ParsableByteArray payload, int packetSequenceNumber)
```

Comment:
I would recommend renaming the method to `validateVp9Descriptor`. And document that the position in the `ParsableByteArray` is moved upon returnning.

We used this javadoc for VP8:

```
  /**
   * Returns {@code true} and sets the {@link ParsableByteArray#getPosition() payload.position} to
   * the end of the descriptor, if a valid VP8 descriptor is present.
   */

```